### ISSUE TITLE
Assigning a T valued Dict to an Optional[T] valued dict is reported as invalid

### ISSUE THREAD

**Issue by haydenjeune** [microsoft/pyright#2171]
**Describe the bug**
Assigning a T valued Dict to an Optional[T] valued Dict is reported as invalid.
It seems that this case is valid, however the reverse should not be.

**To Reproduce**
```python
from typing import Dict, Optional

a: Dict[str, Optional[int]] = {}
b: Dict[str, int] = {}

b = a # this is definitely not valid
a = b # should this be valid?
```

pyright reports both as errors
```
  6:5 - error: Expression of type "Dict[str, int | None]" cannot be assigned to declared type "Dict[str, int]"
  TypeVar "_VT" is invariant
    Type "int | None" cannot be assigned to type "int" (reportGeneralTypeIssues)
  7:5 - error: Expression of type "Dict[str, int]" cannot be assigned to declared type "Dict[str, int | None]"
  TypeVar "_VT" is invariant
    Type "int" cannot be assigned to type "int | None" 
```

**Expected behavior**
pyright continues to report `b = a` as an error, however will allow `a = b`

**VS Code extension or command-line**
Both. Is reproducible with pyright 1.1.161

**Comment by erictraut** [comment]
This is correct behavior defined by PEP 484.

The `dict` class has two type parameters, and both of them are invariant. That means that they must match exactly for type compatibility. This is because dictionaries are mutable. Consider, for example, that `a = b` was allowed, and you followed that assignment with the statement `a["x"] = None`. That would violate the type assumptions of `b`, which cannot contain values of type `None`.

For more details, please refer to [this documentation](https://github.com/microsoft/pyright/blob/main/docs/type-concepts.md#generic-types).

**Comment by erictraut** [comment]
If you replace `Dict` with `Mapping`, it will work as you are expecting. That's because `Mapping` is an immutable form of `Dict`.

**Comment by haydenjeune** [comment]
Ah ok, that makes sense. Thank you!
